File: pathnamejup/pathname/resources/checker/section_1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <pwd.h>
#include <sys/stat.h>
#include <sys/types.h>

// A C file is being used for Section 1, as it contains answers that would be
// visible to the students. When steps are finished, a finish_step_X() function
// is called, which writes out the "answer" or the next set of files.

// This must match the string in the Python file.
#define REQUIRED_SECRET "897dba53d0bd4e72adf9215db3948"

void finish_step_2() {
    // Get the home directory.
    const char *homedir = getenv("HOME");
    if (homedir == NULL) {
        homedir = getpwuid(getuid())->pw_dir;
    }

    // This is to create the templates directory within their home directory.
    char templates_dir[50];
    snprintf(templates_dir, sizeof(templates_dir), "%s/templates", homedir);
    mkdir(templates_dir, 0755);

    // Construct file path for the HTML file.
    char filepath[256];
    snprintf(filepath, sizeof(filepath), "%s/step_3.html", templates_dir);

    // Creating the file for the next step.
    FILE *file = fopen(filepath, "w");
    const char *html_content =
        "<!DOCTYPE html>\n"
        "<html>\n"
        "<head>\n"
        "    <title>Step 3 Demo</title>\n"
        "</head>\n"
        "<body>\n"
        "    <h1>Welcome to Step 3!</h1>\n"
        "    <p>Previously, you printed an HTML tag. Now, you're displaying an entire file.</p>\n"
        "</body>\n"
        "</html>\n";
    fprintf(file, "%s", html_content);
    fclose(file);
}

void finish_step_3() {
    // Create /lab if it doesn't exist.
    struct stat st = {0};
    if (stat("/lab", &st) == -1) {
        // Create the directory.
        if (mkdir("/lab", 0755) == 0) {
            FILE *file = fopen("/lab/memo.py", "w");
            mkdir("/lab/templates", 0755);

            // Basic starter code for /lab/memo.py.
            const char *python_code =
                "from flask import Flask, render_template, request, redirect, url_for, send_file, abort\n"
                "import os\n"
                "\n"
                "# For debugging:\n"
                "import sys\n"
                "\n"
                "app = Flask(__name__)\n"
                "MEMO_DIR = 'memos'\n"
                "\n"
                "os.makedirs(MEMO_DIR, exist_ok=True)\n"
                "\n"
                "@app.route('/')\n"
                "def index():\n"
                "    return render_template('index.html')\n"
                "\n"
                "@app.route('/add_memo', methods=['POST'])\n"
                "def add_memo():\n"
                "    # return (redirect to index())\n"
                "\n"
                "@app.route('/delete_memo/<int:memo_id>', methods=['POST'])\n"
                "def delete_memo(memo_id):\n"
                "    # return (redirect to index())\n"
                "\n"
                "@app.route('/memo/<path:memo_id>')\n"
                "def view_memo(memo_id):\n"
                "    return render_template('view_memo.html')\n";
            fprintf(file, "%s", python_code);
            fclose(file);

            // index.html
            FILE *file2 = fopen("/lab/templates/index.html", "w");
            const char *html_code_1 =
                "<!DOCTYPE html>\n"
                "<html>\n"
                "<head>\n"
                "    <title>Under Construction</title>\n"
                "</head>\n"
                "<body>\n"
                "    <h1>Website Under Construction</h1>\n"
                "    <p>We will be back shortly!</p>\n"
                "</body>\n"
                "</html>\n";
            fprintf(file2, "%s", html_code_1);
            fclose(file2);

            // view_memo.html
            FILE *file3 = fopen("/lab/templates/view_memo.html", "w");
            const char *html_code_2 =
                "<!DOCTYPE html>\n"
                "<html>\n"
                "<head>\n"
                "    <title>Under Construction</title>\n"
                "</head>\n"
                "<body>\n"
                "    <h1>Website Under Construction</h1>\n"
                "    <p>We will be back shortly!</p>\n"
                "</body>\n"
                "</html>\n";
            fprintf(file3, "%s", html_code_2);
            fclose(file3);

        } else {
            perror("mkdir");
        }
    }
}

void finish_step_4() {
    FILE *file1 = fopen("/lab/memo.py", "w");
    const char *python_code = "from flask import Flask, render_template, request, redirect, url_for, send_file, abort\n"
               "import os\n"
               "\n"
               "# For debugging:\n"
               "import sys\n"
               "\n"
               "app = Flask(__name__)\n"
               "MEMO_DIR = 'memos'\n"
               "\n"
               "os.makedirs(MEMO_DIR, exist_ok=True)\n"
               "\n"
               "@app.route('/')\n"
               "def index():\n"
               "    # Get the memos, sort them, then print index.html where \"memos\" is the list of memos.\n"
               "    memos = load_memos()\n"
               "    sorted_memos = dict(sorted(memos.items()))\n"
               "    return render_template('index.html', memos=sorted_memos)\n"
               "\n"
               "@app.route('/add_memo', methods=['POST'])\n"
               "def add_memo():\n"
               "    # Get the form with the name \"memo\", then get its content.\n"
               "    memo_content = request.form.get('memo')\n"
               "    # If it's not empty...\n"
               "    if memo_content:\n"
               "        # Get the next available memo ID and create a path with it.\n"
               "        memo_id = get_next_memo_id()\n"
               "        memo_filename = os.path.join(MEMO_DIR, str(memo_id))\n"
               "        # Create a file of the memo, and store it in memos/.\n"
               "        with open(memo_filename, 'w') as memo_file:\n"
               "            memo_file.write(memo_content)\n"
               "\n"
               "    # Redirect back to the index page.\n"
               "    return redirect(url_for('index'))\n"
               "\n"
               "@app.route('/delete_memo/<int:memo_id>', methods=['POST'])\n"
               "def delete_memo(memo_id):\n"
               "    # TO-DO: Add the functionality here for deleting files.\n"
               "\n"
               "    return redirect(url_for('index'))\n"
               "\n"
               "@app.route('/memo/<path:memo_id>')\n"
               "def view_memo(memo_id):\n"
               "    # Get the URL of the memo.\n"
               "    memo_filename = os.path.join(MEMO_DIR, str(memo_id))\n"
               "    memo_content = \"\"\n"
               "\n"
               "    # Attempt to access the file.\n"
               "    try:\n"
               "        # Read in the file as memo_content.\n"
               "        with open(memo_filename) as file:\n"
               "            memo_content = file.read()\n"
               "    except FileNotFoundError:\n"
               "        abort(404)\n"
               "\n"
               "    # Check if memo_id is numeric. If it's not, an error. Set memo_id = -1.\n"
               "    if not memo_id.isnumeric():\n"
               "        memo_id = -1\n"
               "\n"
               "    # Print view_memo.html page.\n"
               "    return render_template('view_memo.html', memo_id=memo_id, memo=memo_content)\n"
               "\n"
               "def load_memos():\n"
               "    memos = {}\n"
               "    for filename in os.listdir(MEMO_DIR):\n"
               "        with open(os.path.join(MEMO_DIR, filename), 'r') as memo_file:\n"
               "            memos[filename] = memo_file.read()\n"
               "    return memos\n"
               "\n"
               "def get_next_memo_id():\n"
               "    # Create a list of all the files in the memos/ directory.\n"
               "    filenames = os.listdir(MEMO_DIR)\n"
               "    # Create a list of the existing memos.\n"
               "    existing_ids = [int(filename) for filename in filenames if filename.isnumeric()]\n"
               "    # Get the maximum value, add one, then return it. This is the next memo id.\n"
               "    return max(existing_ids, default=0) + 1\n"
               "\n"
               "if __name__ == '__main__':\n"
               "    app.run(debug=True)\n";

    fprintf(file1, "%s", python_code);
    fclose(file1);

    // Creating the HTML for the view_memo.html template.
    FILE *file2 = fopen("/lab/templates/view_memo.html", "w");
    const char *html_code_1 =
        "<!DOCTYPE html>\n"
        "<html lang=\"en\">\n"
        "<head>\n"
        "    <meta charset=\"UTF-8\">\n"
        "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n"
        "    <title>Memo {{ memo_id }}</title>\n"
        "</head>\n"
        "<body>\n"
        "    <h1>Memo {{ memo_id }}</h1>\n"
        "    <p>{{ memo }}</p>\n"
        "    <form action=\"{{ url_for('delete_memo', memo_id=memo_id) }}\" method=\"POST\">\n"
        "        <button type=\"submit\">Delete Memo</button>\n"
        "    </form>\n"
        "    <a href=\"{{ url_for('index') }}\">Back to memos</a>\n"
        "</body>\n"
        "</html>\n";
    fprintf(file2, "%s", html_code_1);
    fclose(file2);

    // Creating the HTML for the index.html template.
    FILE *file3 = fopen("/lab/templates/index.html", "w");
    const char *html_code_2 =
        "<!DOCTYPE html>\n"
        "<html lang=\"en\">\n"
        "<head>\n"
        "    <meta charset=\"UTF-8\">\n"
        "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n"
        "    <title>Memos</title>\n"
        "</head>\n"
        "<body>\n"
        "    <h1>My Memos</h1>\n"
        "    <form action=\"{{ url_for('add_memo') }}\" method=\"POST\">\n"
        "        <textarea name=\"memo\" rows=5 style=\"width: 30%;\" placeholder=\"Enter memo here\"></textarea><br>\n"
        "        <button type=\"submit\">Add Memo</button>\n"
        "    </form>\n"
        "    <h2>Access Memos</h2>\n"
        "    <form>\n"
        "        <select onchange=\"window.location.href=this.value;\">\n"
        "            <option value=\"\">Select a memo</option>\n"
        "            {% for memo_id, memo in memos.items() %}\n"
        "                <option value=\"{{ url_for('view_memo', memo_id=memo_id) }}\">Memo {{ memo_id }}</option>\n"
        "            {% endfor %}\n"
        "        </select>\n"
        "    </form>\n"
        "</body>\n"
        "</html>\n";
    fprintf(file3, "%s", html_code_2);
    fclose(file3);
}

void finish_step_5() {
    FILE *file1 = fopen("/lab/memo.py", "w");
    const char *python_code = "from flask import Flask, render_template, request, redirect, url_for, send_file, abort\n"
               "import os\n"
               "\n"
               "# For debugging:\n"
               "import sys\n"
               "\n"
               "app = Flask(__name__)\n"
               "MEMO_DIR = 'memos'\n"
               "\n"
               "os.makedirs(MEMO_DIR, exist_ok=True)\n"
               "\n"
               "@app.route('/')\n"
               "def index():\n"
               "    # Get the memos, sort them, then print index.html where \"memos\" is the list of memos.\n"
               "    memos = load_memos()\n"
               "    sorted_memos = dict(sorted(memos.items()))\n"
               "    return render_template('index.html', memos=sorted_memos)\n"
               "\n"
               "@app.route('/add_memo', methods=['POST'])\n"
               "def add_memo():\n"
               "    # Get the form with the name \"memo\", then get its content.\n"
               "    memo_content = request.form.get('memo')\n"
               "    # If it's not empty...\n"
               "    if memo_content:\n"
               "        # Get the next available memo ID and create a path with it.\n"
               "        memo_id = get_next_memo_id()\n"
               "        memo_filename = os.path.join(MEMO_DIR, str(memo_id))\n"
               "        # Create a file of the memo, and store it in memos/.\n"
               "        with open(memo_filename, 'w') as memo_file:\n"
               "            memo_file.write(memo_content)\n"
               "\n"
               "    # Redirect back to the index page.\n"
               "    return redirect(url_for('index'))\n"
               "\n"
               "@app.route('/delete_memo/<int:memo_id>', methods=['POST'])\n"
               "def delete_memo(memo_id):\n"
               "    memo_filename = os.path.join(MEMO_DIR, str(memo_id))\n"
               "    if os.path.exists(memo_filename):\n"
               "        os.remove(memo_filename)\n"
               "    return redirect(url_for('index'))\n"
               "\n"
               "@app.route('/memo/<path:memo_id>')\n"
               "def view_memo(memo_id):\n"
               "    # Get the URL of the memo.\n"
               "    memo_filename = os.path.join(MEMO_DIR, str(memo_id))\n"
               "    memo_content = \"\"\n"
               "\n"
               "    ### Step 12 Solution START ###\n"
               "\n"
               "\n"
               "\n"
               "    ### Step 12 Solution END ###\n\n"
               "    # Attempt to access the file.\n"
               "    try:\n"
               "        # Read in the file as memo_content.\n"
               "        with open(memo_filename) as file:\n"
               "            memo_content = file.read()\n"
               "    except FileNotFoundError:\n"
               "        abort(404)\n"
               "\n"
               "    # Check if memo_id is numeric. If it's not, an error. Set memo_id = -1.\n"
               "    if not memo_id.isnumeric():\n"
               "        memo_id = -1\n"
               "\n"
               "    # Print view_memo.html page.\n"
               "    return render_template('view_memo.html', memo_id=memo_id, memo=memo_content)\n"
               "\n"
               "def load_memos():\n"
               "    memos = {}\n"
               "    for filename in os.listdir(MEMO_DIR):\n"
               "        with open(os.path.join(MEMO_DIR, filename), 'r') as memo_file:\n"
               "            memos[filename] = memo_file.read()\n"
               "    return memos\n"
               "\n"
               "def get_next_memo_id():\n"
               "    # Create a list of all the files in the memos/ directory.\n"
               "    filenames = os.listdir(MEMO_DIR)\n"
               "    # Create a list of the existing memos.\n"
               "    existing_ids = [int(filename) for filename in filenames if filename.isnumeric()]\n"
               "    # Get the maximum value, add one, then return it. This is the next memo id.\n"
               "    return max(existing_ids, default=0) + 1\n"
               "\n"
               "if __name__ == '__main__':\n"
               "    app.run(debug=True)\n";

    fprintf(file1, "%s", python_code);
    fclose(file1);
}

int main(int argc, char *argv[])
{
    // Checking the environment variable to make sure that the Python file is calling this file.
    const char* secret = getenv("CHECKER_SECRET");
    if (!secret || strcmp(secret, REQUIRED_SECRET) != 0) {
        fprintf(stderr, "Error: This program may only be run via section_1.py.\n");
        return 3;
    }

    // Checks the usage.
    if (argc != 2) {
        printf("Usage: ./section_1 <step_num>\n");
        return 2;
    }

    // Converts the argument into an int so that we can use it in the switch-case.
    int step = atoi(argv[1]);

    // All of the logic is being done in the Python program. We're just going to generate the next step.
    switch (step) {
        case 2:
            finish_step_2();
            return 0;
        case 3:
            finish_step_3();
            return 0;
        case 4:
            finish_step_4();
            return 0;
        case 5:
            finish_step_5();
            return 0;
        default:
            // Unrecognized step.
            return 1;
    }
}

File: pathnamejup/pathname/resources/checker/section_1.py
#!/usr/bin/python3
import subprocess
import sys
import os
import getpass
import time
import shutil
import re

# This is an environment variable that's used to run ./section_1 while requiring section_1.py to be ran with it.
SECRET_VALUE = "897dba53d0bd4e72adf9215db3948"

def main():
    if len(sys.argv) != 2:
        print("Usage: ./section_1.py <step>")
        sys.exit(2)

    step = sys.argv[1]
    user = getpass.getuser()

    # Kill any leftover process on port 5010.
    result = subprocess.run(['lsof', '-t', '-i:5010'], capture_output=True, text=True)
    pids = result.stdout.strip()
    if pids:
        subprocess.run(['kill', *pids.split()])

    # Creating the environment variable.
    env = os.environ.copy()
    env["CHECKER_SECRET"] = SECRET_VALUE

    # Checking Step 2.
    if step == "2":
        # Checks to see if the student has previously passed this step.
        resp_path = "/home/.checker/responses/step_2_response.txt"
        if os.path.exists(resp_path):
            with open(resp_path, "r") as f:
                old_data = f.read()

            # Checking to make sure that the "Hello, World!" string was printed previously.
            if "<p>Hello, World!</p>" in old_data:
                # Pass with success. Was completed correctly beforehand.
                sys.exit(1)
            else:
                # This shouldn't have happened, as the student cannot previously "pass" with the wrong output.
                # Remove the bad file, then continue with the rest of the check.
                os.remove(resp_path)

        # Continuing with the proper check.
        path_2 = f"/home/{user}/step_2.py"
        # File doesn't exist.
        if not os.path.exists(path_2):
            sys.exit(2)

        # Heading into the student's directory so that we can open a process to run the Flask application.
        os.chdir(f"/home/{user}")

        # Start the student's server on port 5010 so that there's no conflicts.
        process = subprocess.Popen(
            ["python3", "-m", "flask", "--app", "step_2", "run", "-p", "5010"],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
        )
        # To prevent any race conditions.
        time.sleep(2)

        # Calling a curl request and getting the output.
        curl_result = subprocess.run(
            ["curl", "127.0.0.1:5010"],
            capture_output=True, text=True
        )

        # Finish the task so that we can read the output.
        process.terminate()
        process.wait()

        # Checks if "Hello, World!" is printed properly. It should, since the answer is given to them in the notebook.
        if "<p>Hello, World!</p>" in curl_result.stdout:
            # Save a response file so we know we've passed.
            with open(resp_path, "w") as f:
                f.write(curl_result.stdout)

            # Before exiting successfully, create the next file for ~/step_3.py.
            c_result = subprocess.run(["/home/.checker/section_1", "2"], env=env, capture_output=True)

            # Should be successful.
            if c_result.returncode == 1:
                # If they pass, copy step_2.py to step_3.py for the next step.
                path_3 = f"/home/{user}/step_3.py"
                if not os.path.exists(path_3):
                    shutil.copyfile(path_2, path_3)
                sys.exit(1)

            # Something odd happened. They passed, but something is wrong with the checker script. Return a code of 3.
            else:
                os.remove(resp_path)
                sys.exit(3)
        else:
            # They have the wrong output (shouldn't happen, as the answer is given to them in the step).
            sys.exit(0)

    # Checking Step 3.
    elif step == "3":
        # Getting the response path, so that we can check if it was previously passed.
        resp_path = "/home/.checker/responses/step_3_response.txt"
    
        # If this file already exists, check if we previously passed Step 3.
        if os.path.exists(resp_path):
            with open(resp_path, "r") as f:
                old_data = f.read()
            # Checking the response and seeing if the template was printed.
            if "<h1>Welcome to Step 3!</h1>" in old_data:
                # Previously passed. Skip the step.
                sys.exit(1)
            else:
                # This was an invalid response that somehow got passed. Remove and continue with the rest of the check.
                os.remove(resp_path)
    
        # Defining a fresh path that we will be testing this on.
        path_3 = f"/home/{user}/step_3.py"

        # In case they haven't made step_3.py yet. This should've been done automatically.
        if not os.path.exists(path_3):
            sys.exit(2)
    
        # Start the student's server, just like Step 2. (This will be repeated throughout each step of the file).
        os.chdir(f"/home/{user}")
        process = subprocess.Popen(
            ["python3", "-m", "flask", "--app", "step_3", "run", "-p", "5010"],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
        )
        time.sleep(2)
    
        curl_result = subprocess.run(
            ["curl", "127.0.0.1:5010"],
            capture_output=True, text=True
        )
    
        process.terminate()
        process.wait()
    
        # Now check if the returned HTML contains <h1>Welcome to Step 3!</h1>.
        if "<h1>Welcome to Step 3!</h1>" in curl_result.stdout:
            # Save a response file so we know we've passed.
            with open(resp_path, "w") as f:
                # Storing the output here.
                f.write(curl_result.stdout)
    
            # Now call the C code to finalize Step 3. sudo is required here since we need to make a directory in /.
            c_result = subprocess.run(["sudo", "-E", "/home/.checker/section_1", "3"], env=env, capture_output=True)

            # Should work.
            if c_result.returncode == 1:
                # Giving ownership to the student so that they can freely work within it.
                if os.path.exists("/lab"):
                    st_info = os.stat("/lab")
                    if st_info.st_uid == 0:
                        subprocess.run(["sudo", "chown", "-R", user, "/lab"])
                sys.exit(1)

            # Something went wrong with the checker script.
            else:
                # If the C code fails, remove the response so we can try again later.
                os.remove(resp_path)
                sys.exit(3)
        else:
            # The student is still returning "<p>Hello, World!</p>" or otherwise incorrect.
            sys.exit(0)

    # Checking Step 4.
    elif step == "4":
        # Defining the path, like always.
        resp_path = "/home/.checker/responses/step_4_response.txt"

        # If step_4_response.txt already exists, check if it was successful originally.
        if os.path.exists(resp_path):
            with open(resp_path, "r") as f:
                old_data = f.read()

            # There should've been a divider here, which we will split.
            if "\n-DIVIDER-\n" in old_data:
                parts = old_data.split("\n-DIVIDER-\n", 1)
                old_code = parts[0]
                old_curl = parts[1]

                # Defined the correct response, with some leniency in the answer.
                pattern = r'return\s+redirect\s*\(\s*url_for\s*\(\s*[\'"]index[\'"]\s*\)\s*\)'
                matches = re.findall(pattern, old_code)

                # If the redirect worked, this should be included in the curl request.
                has_redirect_html = ("<p>You should be redirected automatically to the target URL: <a href=\"/\">/</a>. If not, click the link." in old_curl)

                # There should be two matches for the redirects.
                if len(matches) >= 2 and has_redirect_html:
                    sys.exit(1)

                # Their response is actually invalid, but something allowed them to pass. Remove their response.
                else:
                    os.remove(resp_path)

            # If there was no divider, then this file was incorrectly made.
            else:
                os.remove(resp_path)

        # This file should exist.
        if not os.path.exists("/lab/memo.py"):
            sys.exit(2)

        # Read in their file.
        with open("/lab/memo.py", "r") as f:
            user_code = f.read()

        # Creating a regular expression with their response, adding some whitespace and quotation leniency.
        pattern = r'return\s+redirect\s*\(\s*url_for\s*\(\s*[\'"]index[\'"]\s*\)\s*\)'
        matches = re.findall(pattern, user_code)

        # This is in case they forgot to add two of the redirects.
        if len(matches) < 2:
            sys.exit(0)

        # Navigating into their /lab directory so that we can run their Flask application, like the previous steps.
        os.chdir("/lab")
        process = subprocess.Popen(
            ["python3", "-m", "flask", "--app", "memo", "run", "-p", "5010"],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
        )
        time.sleep(2)

        # Attempting to add a memo using a cURL request.
        curl_result = subprocess.run(
            "curl -X POST -d \"memo=test\" -v http://127.0.0.1:5010/add_memo",
            shell=True, capture_output=True, text=True
        )

        process.terminate()
        process.wait()

        # The redirect should be working.
        redirect_snippet = "<p>You should be redirected automatically to the target URL: <a href=\"/\">/</a>. If not, click the link."
        if redirect_snippet in curl_result.stdout:
            # They were successful, and we will need to create a response text file so that we can refer back to it later.
            with open(resp_path, "w") as f:
                f.write(user_code)
                # Need a divider so that we can split the code and the cURL request in the check.
                f.write("\n-DIVIDER-\n")
                f.write(curl_result.stdout)

            # Generating the next step.
            c_result = subprocess.run(["/home/.checker/section_1", "4"], env=env, capture_output=True)

            # This should be working.
            if c_result.returncode == 1:
                sys.exit(1)

            # Otherwise, something went wrong with the checker file.
            else:
                os.remove(resp_path)
                sys.exit(3)
        else:
            sys.exit(0)

    # Checking Step 5.
    elif step == "5":
        resp_path = "/home/.checker/responses/step_5_response.txt"

        # Checking if they previously passed this.
        if os.path.exists(resp_path):
            with open(resp_path, "r") as f:
                old_data = f.read()

            # There should be a divider in their response.
            if "\n-DIVIDER-\n" in old_data:
                parts = old_data.split("\n-DIVIDER-\n", 1)
                old_code = parts[0]
                old_curl = parts[1]

                # This cURL response should be in their correct file.
                if "< HTTP/1.1 302 FOUND" in old_curl:
                    sys.exit(1)

                # Otherwise, they somehow passed, but their notebook generated the wrong file.
                else:
                    os.remove(resp_path)
            
            # Otherwise, remove their response, since this is an invalid file.
            else:
                os.remove(resp_path)

        # Checks to make sure that their memo.py file still exists.
        if not os.path.exists("/lab/memo.py"):
            sys.exit(2)

        # Starting up their Flask application with a process.
        os.chdir("/lab")
        with open("memo.py", "r") as f:
            user_code = f.read()

        process = subprocess.Popen(
            ["python3", "-m", "flask", "--app", "memo", "run", "-p", "5010"],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
        )
        time.sleep(2)

        # Create a test memo with 9999 as the ID.
        test_memo_path = "/lab/memos/9999"
        with open(test_memo_path, "w") as f:
            f.write("test")

        # Attempt to delete it with their function.
        curl_result = subprocess.run(
            "curl -X POST -v http://127.0.0.1:5010/delete_memo/9999",
            shell=True, capture_output=True, text=True
        )

        # There was no valid delete function.
        if "< HTTP/1.1 302 FOUND" not in curl_result.stderr:
            process.terminate()
            process.wait()
            if os.path.exists(test_memo_path):
                os.remove(test_memo_path)
            sys.exit(0)

        # This shouldn't happen, since this was automatically generated for them.
        if os.path.exists(test_memo_path):
            process.terminate()
            process.wait()
            sys.exit(0)

        # At this point, they should've passed the step. Create their "success" file.
        with open(resp_path, "w") as f:
            f.write(user_code)
            f.write("\n-DIVIDER-\n")
            f.write(curl_result.stderr.strip())

        # Generate the file for their next step (which is Step 11).
        c_result = subprocess.run(["/home/.checker/section_1", "5"], env=env, capture_output=True)

        # Suspending the process.
        process.terminate()
        process.wait()

        # If the next step was correctly generated, exit successfully.
        if c_result.returncode == 1:
            sys.exit(1)

        # This will only happen if an error happened with the checker script.
        else:
            os.remove(resp_path)
            sys.exit(3)

    # Invalid parameter(s).
    else:
        sys.exit(2)

if __name__ == "__main__":
    main()
